PROM device data

EmilioPeJu opened the pull request:
Extract device information from PROM memory

The following information is obtained:
- Device node name for the mmap-able device to access registers
- Each one of the DMA memory regions

This change also includes unit tests and a script to facilitate creating
the PROM data based on a configuration file

EmilioPeJu committed:
Extract device information from PROM memory

The following information is obtained:
- Device node name for the mmap-able device to access registers
- Each one of the DMA memory regions

This change also includes unit tests and a script to facilitate creating
the PROM data based on a configuration file

EmilioPeJu committed:
Improve error handling for DMA control

- Don't ignore reset error
- Log a message when a DMA decoding, interrupt or slave problem occurs

EmilioPeJu committed:
Create sysfs attributes to read prom and prom used

The first attribute reads the prom via PCI while the second one reads it
from the buffer that is used to parse it and therefore it represents what
the driver is using

Araneidae commented on driver/dma_control.c:
```diff
@@ -83,38 +83,54 @@ struct dma_control {
     int count = 0;
-    while (readl(&dma->regs->cdmacr) & CDMACR_Reset  &&
+    int reset_stat = 0;
+    while ((reset_stat=readl(&dma->regs->cdmacr) & CDMACR_Reset)  &&
            time_before(jiffies, deadline))
         count += 1;
```
I don't like assignment expressions at all.  Of course, in this case I agree that this is better than a `while(true) { ... if(...)break; }` loop.  In this particular case I like to use the following trick:
```c
    while (
        reset_stat = readl(&dma->regs->cdmacr),
        (reset_stat & CDMACR_Reset)  &&  time_before(jiffies, deadline))
    {
        count += 1;
    }
```
Here we use the comma operator to separate assignments from the test, and I've used brackets for the one-line body simply for layout.

EmilioPeJu replied:
Done

Araneidae commented on driver/amc_pci_core.c:
```diff
@@ -125,19 +153,35 @@ static int amc_pci_open(struct inode *inode, struct file *file)
-        case MINOR_DDR1:
-            rc = amc_pci_dma_open(file, amc_priv->dma, DDR1_BASE, DDR1_LENGTH);
-            break;
+        if (pentry->tag == PROM_DEVICE_TAG)
```
Think a switch statement on `pentry->tag` would be nice.  I also note we silently ignore unrecognised tags; is that the right behaviour?

EmilioPeJu replied:
Done, although there is a little detail to consider, in C99, we get an error if a declaration happens immediately after a tag (which is used for each case), so I added a dummy ; statement to bypass that

Araneidae replied:
Ah.  Don't do declarations inside switch statements without an enclosing `{ ... }` block; the results are otherwise really rather ugly.

EmilioPeJu replied:
fixed

Araneidae commented on driver/amc_pci_core.c:
```diff
@@ -160,15 +204,45 @@ static int create_device_nodes(
+    prom_for_each_entry(pentry, amc_priv->prom)
+    {
+        // it is assumed that the device entry is the first found
+        if (pentry->tag == PROM_DEVICE_TAG)
```
Again, use a switch here.

EmilioPeJu replied:
Same as other one

Araneidae commented on driver/amc_pci_core.c:
```diff
@@ -160,15 +204,45 @@ static int create_device_nodes(
+        {
+            struct prom_device_entry *device_entry =
+                (struct prom_device_entry *) pentry;
+            device_name = device_entry->name;
```
I guess need to check that `device_name` isn't already assigned?

EmilioPeJu replied:
Done

Araneidae commented on driver/amc_pci_core.c:
```diff
@@ -180,7 +254,7 @@ static void destroy_device_nodes(
     int minor = amc_priv->minor;
 
-    for (int i = 0; i < MINORS_PER_BOARD; i ++)
+    for (int i = 0; i < amc_priv->prom->nentries; i ++)
```
Are we sure this won't try to destroy non-existent device nodes if we ignored some entries?

EmilioPeJu replied:
The probe function will error if there is an unknown entry, therefore the remove function will not be called in that case

Araneidae commented on driver/amc_pci_core.c:
```diff
@@ -274,6 +347,17 @@ static int initialise_board(struct pci_dev *pdev, struct amc_pci *amc_priv)
+    struct prom_context *prom_context = load_prom(amc_priv->ctrl_memory + PROM_OFFSET);
+    if (IS_ERR(prom_context)) {
+        rc = PTR_ERR(prom_context);
+        goto prom_error;
```
Do we need to log an error here?  (Or did `load_prom` already do that for us?)

Araneidae replied:
No, already logged.

Araneidae commented on driver/default_prom.c:
```diff
@@ -0,0 +1,10 @@
+const size_t default_prom_nentries = 3;
```
Would be nice to autogenerate this entire file from the appropriate configuration file instead.

EmilioPeJu replied:
Done, adding auto-generation has been a bit tricky, I had to add it to Kbuild in order to also consider it for dkms

Araneidae commented on driver/dma_control.c:
```diff
@@ -83,38 +83,54 @@ struct dma_control {
 
 
-static void maybe_reset_dma(struct dma_control *dma)
+static bool check_dma_status(struct dma_control *dma)
```
I'm in two minds about `bool` vs `int` for error codes.  The problem is, we want to be consistent ... and most functions seem to be returning an error code.  Not sure it's possible to be fully consistent, to be honest.  Maybe the fact that we're logging an error is a hint that this should be an error code?

EmilioPeJu replied:
Done

Araneidae commented on driver/dma_control.c:
```diff
@@ -158,9 +175,11 @@ ssize_t read_dma_memory(
     /* Reset the DMA engine if necessary. */
-    maybe_reset_dma(dma);
+    TEST_RC(rc = (ssize_t) maybe_reset_dma(dma), reset_error,
+        "Fail to reset DMA");
```
"Failed"

EmilioPeJu replied:
Fixed

Araneidae commented on driver/dma_control.c:
```diff
@@ -235,11 +258,12 @@ int initialise_dma_control(
     init_completion(&dma->dma_done);
 
-    reset_dma_controller(dma);
+    TEST_RC(rc=reset_dma_controller(dma), reset_error, "Fail to reset DMA");
```
No: please don't use assignment expressions unless forced to!  In this case it's easy to write:
```c
rc = reset_dma_controller(dma);
TEST_RC(rc, reset_error, "Fail to reset DMA");
```

Araneidae replied:
Not sure I'm being consistent here, to be honest; the `TEST_...` macros in my other bodies of code (particularly [PandaBlocks-server](https://github.com/PandABlocks/PandABlocks-server) and [epics_device](https://github.com/Araneidae/epics_device) rely on assignment expressions, but we're not using them anywhere else in this driver so I'll stand by this.

EmilioPeJu replied:
I already spitted it, though I can roll that back if you prefer so

Araneidae commented on driver/amc_pci_core.c:
```diff
@@ -160,15 +204,45 @@ static int create_device_nodes(
+    }
     return 0;
 
+prom_error:
```
Need to keep track of which devices have been created and delete the ones that have been on error.

Araneidae commented on driver/amc_pci_core.c:
```diff
@@ -33,7 +36,6 @@ MODULE_VERSION(S(VERSION));
-
 /* Physical layout in DDR address space of the two memory areas. */
 #define DDR0_BASE       0
 #define DDR0_LENGTH     0x80000000      // 2GB
```
Can delete these defaults now!

EmilioPeJu replied:
Done

Araneidae commented on driver/prom_processing.c:
```diff
@@ -0,0 +1,118 @@
+    // on a putative end marker
+    TEST_OK(
+        ent_i < PROM_MAX_LENGTH - PROM_END_ENTRY_SIZE
+            && context->buff[ent_i] == PROM_END_TAG,
```
You never validate the length of the end entry!

EmilioPeJu replied:
Done

EmilioPeJu commented on driver/amc_pci.spec.in:
```diff
@@ -15,31 +15,41 @@ Packager: Michael Abbott <michael.abbott@diamond.ac.uk>
+install -m 0644 %{_sourcedir}/prom_processing.h          %{buildroot}%{dkmsdir}
+install -m 0644 %{_sourcedir}/default_prom.config        %{buildroot}%{dkmsdir}
+install -d                                               %{buildroot}%{dkmsdir}/tools
+install -m 0755 %{_sourcedir}/tools/prom_data_creator.py %{buildroot}%{dkmsdir}/tools
```
I know this line is bigger than 79 but I didn't know how to wrap it in a good way

EmilioPeJu commented on driver/prom_processing.c:
```diff
@@ -0,0 +1,130 @@
+    if (off > PROM_MAX_LENGTH)
+        return -EIO;
+    // offset and count should be multiple of 4 bytes
+    loff_t off_al = off/4*4;
```
This requirement is added to simplify code

lyodaom commented:
Just let you know that I updated the RPM spec to consider these changes and tested it in the AMC720

Araneidae commented on driver/amc_pci_core.c:
```diff
@@ -125,19 +138,39 @@ static int amc_pci_open(struct inode *inode, struct file *file)
-            rc = amc_pci_dma_open(file, amc_priv->dma, DDR0_BASE, DDR0_LENGTH);
+        switch (pentry->tag)
+        {
+        case PROM_DEVICE_TAG:
```
Interesting question.  Do we indent the case to the switch or one level deeper?  Check the rest of the code base, need to be consistent about this.

I always indent deeper:
```c
switch (blah)
{
    case label:
        body;
        break;
    case label2:
    {
        body with braces;
        break;
    }
}
```

EmilioPeJu replied:
done

Araneidae committed:
Change class name to amc_pci and update udev rules accordingly

Also change name in code from DEVICE_NAME to CLASS_NAME.  We don't actually
use the device name in any reporting.

Araneidae committed:
Whitespace changes

Put spaces around = operator, make sure no lines are longer than 80 chars
